File: 3sem/4lab/Banks/Exceptions/InvalidBankAccountOperationException.cs
namespace Banks.Exceptions;

public class InvalidBankAccountOperationException : BanksDomainException
{
    private InvalidBankAccountOperationException(string? message)
        : base(message) { }

    public static InvalidBankAccountOperationException OnInvalidMoneyAmount()
        => new InvalidBankAccountOperationException("Money amount must be positive");

    public static InvalidBankAccountOperationException OnInvalidInterestRate()
        => new InvalidBankAccountOperationException($"Interest rate must be in [0; 1]");

    public static InvalidBankAccountOperationException OnInvalidCreditLimit()
        => new InvalidBankAccountOperationException($"Credit limit must be positive");

    public static InvalidBankAccountOperationException OnInvalidCommission()
        => new InvalidBankAccountOperationException($"Commission must be positive");

    public static InvalidBankAccountOperationException OnShortageOfMoney(decimal sum, decimal balance)
        => new InvalidBankAccountOperationException($"Unable to withdraw {sum}, account has only {balance}");
}

File: 4sem/1lab/src/main/java/BanksRemake/Domain/Exceptions/InvalidBankAccountOperationException.java
package banksRemake.domain.exceptions;

public class InvalidBankAccountOperationException extends BanksRemakeException {
    private InvalidBankAccountOperationException(String message) {
        super(message);
    }

    public static InvalidBankAccountOperationException onInvalidMoneyAmount() {
        return new InvalidBankAccountOperationException("Money amount must be positive");
    }

    public static InvalidBankAccountOperationException onInvalidInterestRate() {
        return new InvalidBankAccountOperationException("Interest rate must be in [0; 1]");
    }

    public static InvalidBankAccountOperationException onInvalidCreditLimit() {
        return new InvalidBankAccountOperationException("Credit limit must be positive");
    }

    public static InvalidBankAccountOperationException onInvalidCommission() {
        return new InvalidBankAccountOperationException("Commission must be positive");
    }

    public static InvalidBankAccountOperationException onShortageOfMoney(double sum, double balance) {
        return new InvalidBankAccountOperationException(String.format("Unable to withdraw %1, account has only %2", sum, balance));
    }
}
